### ISSUE TITLE
fix leader->validator startup delay

### ISSUE THREAD

**Issue by rob-solana** [solana-labs/solana#2317]
#### Problem
Currently, when a node transitions from a leader->validator the newly started validator must read the entire ledger to get back in sync because the "leader bank" is discarded, considered "corrupt" because transactions have been played into it that might be beyond the current leader period.

#### Proposed Solution
Explore options:
1. checkpoint() and collapse() the TPU bank at every tick, and on leader->validator transition pass the leader bank's last checkpoint over to the TVU (implies ownership changes of Bank/Bank_state or modality of the Bank)
2. Run 2 banks: the TPU runs a separate bank and sends entries over to an "always on" TVU.   This option may be better/more correct given fork selection code.  Implies separate identities for leader/validator?

**Comment by sagar-solana** [comment]
It's not as cumbersome as it sounds actually. We're planning on keeping tvu running all the time and everytime a node becomes a leader, it'll use a checkpointed copy of the bank from the tvu. All ledger maintenance will be passed on to the tvu only and tpu only processes txs and broadcasts to the network (and it's local tvu) nothing else.

The problem with rolling back the leader bank is that it's almost never in a good state to reuse for tvu. Txs are in flight in parallel and not necessarily ordered. Maybe @rob-solana can explain better.

**Comment by garious** [comment]
The problems are memory and performance, not complexity. We can't duplicate the bank state or the calls to process_transactions.

**Comment by sagar-solana** [comment]
Fair enough, although I don't think it's as bad as it sounds. The accounts data structure and last ids is all that gets copied over. So we won't be processing txs again but yeah the memory complexity could be a problem. In any case it cannot be worse than it is today with the whole ledger being replayed, haha. Maybe the three of us should discuss this before I get too deep into an implementation :D

**Comment by garious** [comment]
Seems like there's not really 2 options here, and that the problem is the use of VecDeque to store a chain of rollback points. The bank states need to be a DAG, not a list.  The TPU should point to a different entrypoint of the DAG than the TVU, but both should point to the same confirmed state.

**Comment by rob-solana** [comment]
yep, we'd need a hashmap of checkpoints, my initial pass assumed we'd rollback and replay on a single fork at a time

**Comment by garious** [comment]
Yes, there could be a hashmap from PoH hashes to entrypoints into the DAG.
